proposals/imporved-hash-module.md

```diff
@@ -0,0 +1,134 @@
+# Improve hash module
+
+Current implementation of the hash module in standard library reflex the implementation of the Python hash module, which by itself is a good idea, but it has some flaws, which we should correct in Mojo.
+
+# Flaws of the `Hashable` trait
+
+The `Hashable` trait is designed as following:
+
+```mojo
+trait Hashable:
+    fn __hash__(self) -> Int:
+        ...
+```
+
+Which implies that a developer, who writes a new hashable struct needs to return an `Int`. 
+- This API does not provide guidance to how this `Int` value needs to be computed
+- It is impossible to exchange hashing algorithms on demand
+- `Int` as a type is variable length (based on the CPU arch), which might cause issues
+- Such trait design follows the call return principle which, when applied on complex types, will lead to unnecessary computations and memory allocations
+
+## Example
+
+```mojo
+@value
+struct Person(Hashable):
+    var name: String
+    var age: UInt8
+    var friends_names: List[String]
+
+    fn __hash__(self) -> Int:
+        var hashes = List[Int]()
+        hashes.append(hash(self.name))
+        hashes.append(hash(self.age))
+        for friend in self.friends_names:
+            hashes.append(hash(friend[]))
+        
+        # How to combine a hash of hashes ???
+```
+As you can see above we, computed hashes for all of the struct fields, but we are uncertain how to combine those values in a way which produces a good (non compromised) hash value.
+
```

Comment:
I believe to be fair, we should add an example of what the computing of the hash would look like if mojo evolves without this proposal. By following https://docs.python.org/3/reference/datamodel.html#object.__hash__ it would look like this

```mojo
@value
struct Person(Hashable):
    var name: String
    var age: UInt8
    var friends_names: List[String]
    fn __hash__(self) -> Int:
        var final_tuple = (self.name, self.age) + tuple(self.friends)
        return hash(final_tuple)
```
And this has a lot of flaw(lots of copying, possible memory allocation, unpredictable types possible for the tuple, use of the `object` struct, a struct must know how to hash it's members, etc...) but at least the proposal gives a fair chance to the "python's way" of doing the hashing.

Reply:
I mentioned the "python's way" in the proposal.